service/history/shard/context_impl.go

```diff
@@ -98,18 +127,24 @@ const (
 )
 
 func (s *ContextImpl) GetShardID() int32 {
+	// constant from initialization, no need for locks
 	return s.shardID
 }
 
 func (s *ContextImpl) GetService() resource.Resource {
+	// constant from initialization, no need for locks
 	return s.Resource
 }
 
 func (s *ContextImpl) GetExecutionManager() persistence.ExecutionManager {
+	// constant from initialization, no need for locks
 	return s.executionManager
 }
 
 func (s *ContextImpl) GetEngine() Engine {
+	s.rLock()
+	defer s.rUnlock()
+
 	return s.engine
```

Comment:
why locking here?

Reply:
it is possible that engine can be nil? get engine while shard is being shut down.
can returning a nil engine causing NPE?